sbws/lib/relayprioritizer.py

```diff
@@ -23,71 +22,100 @@ def __init__(self, args, conf, relay_list, result_dump):
+        Since there could be new relays in the network while measuring the
+        list of relays returned by this method, this method is run again
+        before all the relays in the network are measured.
+
+        .. note::
+
+            In a future refactor, instead of having a static fraction of relays
+            to be measured, this method could be call when it's known that
+            there're X number of new relays in the network.
+
+        Since measurements made before than X days ago (too old) are not
+        considered, and the initial list of past measurements is only filtered
+        when the scanner starts, it's needed to filter here again to discard
+        those measurements.
+
+        :param bool prioritize_result_error: whether prioritize or not
+            measurements that did not succed.
+        :param bool return_fraction: whether to return only a fraction of the
+            relays seen in the network or return all.
+        return: a generator of the new ordered list of relays to measure next.
+
+        """
         fn_tstart = Decimal(time.time())
         relays = set(copy.deepcopy(self.relay_list.relays))
         if not self.measure_authorities:
             relays = relays.difference(set(self.relay_list.authorities))
+        # Since there will be new measurements every time this method is called
+        # again, update the list of results.
+        # In a future refactor with other data structure there should not be
+        # needed.
         rd = self.result_dump
         for relay in relays:
             results = rd.results_for_relay(relay)
             priority = 0
             # The time before which we do not consider results valid anymore
             oldest_allowed = time.time() - self.fresh_seconds
             for result in results:
-                assert isinstance(result, Result)
                 # Ignore results that are too far in the past
                 if result.time < oldest_allowed:
                     continue
                 # Calculate freshness as the remaining time until this result
                 # is no longer valid
                 freshness = result.time - oldest_allowed
-                if isinstance(result, ResultError):
-                    # Reduce the freshness for results containing errors so
-                    # that they are not de-prioritized as much. This way, we
-                    # will come back to them sooner to try again.
-                    assert result.freshness_reduction_factor >= 0.0
-                    assert result.freshness_reduction_factor <= 1.0
-                    log.debug('Cutting freshness for a %s result by %d%% for '
-                              '%s', result.type.value,
-                              result.freshness_reduction_factor * 100,
-                              relay.nickname)
+                if isinstance(result, ResultError) \
+                        and prioritize_result_error is True:
```

Comment:
I think this one can just be `prioritize_result_error` and you can leave out the `is True`?

Reply:
good point, i think that to avoid errors caused by python not being statically typed, is better to check the type when there's ambiguity.
For instance, if at some point we want to prioritize relays that have more than 3 errors (and we reuse the same argument or in the same position because of a human mistake), and we pass to the method `prioritize_result_error=2`, the code would be executed, when it's not what was intended. Makes sense?